apis/apps/v1alpha1/cluster_conversion.go

```diff
@@ -0,0 +1,247 @@
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/jinzhu/copier"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	appsv1 "github.com/apecloud/kubeblocks/apis/apps/v1"
+	workloads "github.com/apecloud/kubeblocks/apis/workloads/v1alpha1"
+)
+
+// ConvertTo converts this Cluster to the Hub version (v1).
+func (r *Cluster) ConvertTo(dstRaw conversion.Hub) error {
+	dst := dstRaw.(*appsv1.Cluster)
+
+	// objectMeta
+	dst.ObjectMeta = r.ObjectMeta
+
+	// spec
+	if err := copier.Copy(&dst.Spec, &r.Spec); err != nil {
+		return err
+	}
+	if err := incrementConvertTo(r, dst); err != nil {
+		return err
+	}
+
+	// status
+	if err := copier.Copy(&dst.Status, &r.Status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConvertFrom converts from the Hub version (v1) to this version.
+func (r *Cluster) ConvertFrom(srcRaw conversion.Hub) error {
+	src := srcRaw.(*appsv1.Cluster)
+
+	// objectMeta
+	r.ObjectMeta = src.ObjectMeta
+
+	// spec
+	if err := copier.Copy(&r.Spec, &src.Spec); err != nil {
+		return err
+	}
+	if err := incrementConvertFrom(r, src, &clusterConverter{}); err != nil {
+		return err
+	}
+
+	// status
+	if err := copier.Copy(&r.Status, &src.Status); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Cluster) incrementConvertTo(dstRaw metav1.Object) (incrementChange, error) {
+	// changed
+	r.changesToCluster(dstRaw.(*appsv1.Cluster))
+
+	// deleted
+	c := &clusterConverter{}
+	c.fromCluster(r)
+	return c, nil
+}
+
+func (r *Cluster) incrementConvertFrom(srcRaw metav1.Object, ic incrementChange) error {
+	// deleted
+	c := ic.(*clusterConverter)
+	c.toCluster(r)
+
+	// changed
+	r.changesFromCluster(srcRaw.(*appsv1.Cluster))
+
+	return nil
+}
+
+func (r *Cluster) changesToCluster(cluster *appsv1.Cluster) {
+	// changed:
+	//   spec
+	//     components
+	//       - volumeClaimTemplates
+	//           spec:
+	//             resources: corev1.ResourceRequirements -> corev1.VolumeResourceRequirements
+	//         podUpdatePolicy: *workloads.PodUpdatePolicyType -> *PodUpdatePolicyType
+	//     sharings
+	//       - template
+	//           volumeClaimTemplates
+	//             spec:
+	//               resources: corev1.ResourceRequirements -> corev1.VolumeResourceRequirements
+	//           podUpdatePolicy: *workloads.PodUpdatePolicyType -> *PodUpdatePolicyType
+	//   status
+	//     components
+	//       - message: ComponentMessageMap -> map[string]string
```

Comment:
to be implemented?

Reply:
The type names of all these fields have changed, but the actual data types are the same, so they can be handled through the json serialization and deserialization.

Reply:
Ok, but I got confused:

```golang
func (r *Component) incrementConvertTo(dstRaw metav1.Object) (incrementChange, error) {
	// changed
	comp := dstRaw.(*appsv1.Component)
	comp.Status.Message = r.Status.Message  // <-- why this field need to be copied by hand?

	// deleted
	return &componentConverter{
		Affinity:               r.Spec.Affinity,
		Tolerations:            r.Spec.Tolerations,
		InstanceUpdateStrategy: r.Spec.InstanceUpdateStrategy,
	}, nil
}
```

Reply:
I just added it on the fly

Reply:
Otherwise I would have to comments them somewhere 😂